src/resolver.ts

```diff
@@ -2022,7 +2022,7 @@ export class Resolver extends DiagnosticEmitter {
             if (overload) return overload.signature.returnType;
           }
         }
-        return leftType == Type.f32 ? Type.f32 : Type.f64;
+        return leftType;
```

Comment:
This seems to be missing that there's sometimes a `commonDenominator` computation leading to a different result than `leftType`.

Reply:
so that better to do in this case?

Reply:
One can think of the resolver as an oracle that tells us what type the compiler will yield on compile. So, if `commonDenominator` is computed and used in the compiler, the same type must be computed and returned by the resolver as well. Take for example `someI32 ** someF64` which will compile to an `f64`, but the resolver will oracle that the type is `i32`, which is wrong it will yield to issues where for example `I32#toString` is called with an `f64`.

Reply:
Are you planning to look into this as well? Would need something like resolving both left and right, and performing the same common denominator computation as [done here](https://github.com/AssemblyScript/assemblyscript/pull/1146/files#diff-d5edbe17e79bd7d0417b0bac7e92be06R4758). The old resolve code is fine for `f32` and `f64` I guess, but the resolved type of integer pow now is the common type.

Reply:
I thought this already solved? I use different types in tests and it seems all work as expected:
https://github.com/AssemblyScript/assemblyscript/pull/1146/files#diff-d0b06574e5eed6802cbc0a204227084dR4114

https://github.com/AssemblyScript/assemblyscript/pull/1146/files#diff-d0b06574e5eed6802cbc0a204227084dR4122

Reply:
Or I missed something? Could you point where I should pay attention?

Reply:
What typically happens there is that if there's just a `a ** b`, the compiler will compile it right away. Expressions can be arbitrarily complex, however, or have type inferring contexts, where sometimes the resolver is invoked to determine a sub expressions's type before compiling it. This code path is not currently covered in this PR, but to cover these cases, the resolver must yield the exact type the compiler would compile to.

In this specific case, there should be a common denominator computation in the resolver as well where the compiler does it. So, instead of `return leftType` here, if `leftType´ is not a float, `right` should become resolved and the common type returned. Implementing this is typically easy by looking at the respective compile bits, and making the resolve code behave just like it.

Reply:
So I should use `let commonType = Type.commonDenominator(leftType, rightType, false);` and return `commonType` instead `leftType`? And perhaps remove `commonDenominator` from code relate to power expression in module.ts?

Reply:
It's essentially this piece of logic that must be performed by the resolver as well:

```ts
        if (compound) {
          leftExpr = this.ensureSmallIntegerWrap(leftExpr, leftType);
          rightExpr = this.compileExpression(right, leftType, Constraints.CONV_IMPLICIT);
          rightType = commonType = this.currentType;
        } else {
          rightExpr = this.compileExpression(right, leftType);
          rightType = this.currentType;
          commonType = Type.commonDenominator(leftType, rightType, false);
          if (commonType) {
            leftExpr = this.convertExpression(leftExpr,
              leftType, commonType,
              false, true, // !
              left
            );
            leftType = commonType;
            rightExpr = this.convertExpression(rightExpr,
              rightType, commonType,
              false, true, // !
              right
            );
            rightType = commonType;
          } else {
            this.error(
              DiagnosticCode.Operator_0_cannot_be_applied_to_types_1_and_2,
              expression.range, "**", leftType.toString(), rightType.toString()
            );
            this.currentType = contextualType;
            return module.unreachable();
          }
        }
```

i.e. compute the common denominator if the expression is not compound, and return `commonType` instead of just `leftType`. Something like this:

```ts
        if (compound) {
          return leftType; // what we did before
        } else {
          let rightType = this.resolveExpression(right, ctxFlow, leftType, reportMode);
          let commonType = Type.commonDenominator(leftType, rightType, false);
          if (commonType) {
            return commonType;
          } else {
            if (reportMode == ReportMode.REPORT) {
              this.error(
                DiagnosticCode.Operator_0_cannot_be_applied_to_types_1_and_2,
                expression.range, "**", leftType.toString(), rightType.toString()
              );
            }
            return ctxType;
          }
        }
```